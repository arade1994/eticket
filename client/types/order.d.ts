import { type Ticket } from "./ticket";

export interface Order {
  id: string;
  userId: string;
  status: OrderStatus;
  expiresAt: Date;
  version: number;
  ticket: Ticket;
}

export declare enum OrderStatus {
  Created = "created",
  Cancelled = "cancelled",
  AwaitingPayment = "awaiting:payment",
  Complete = "complete",
}
