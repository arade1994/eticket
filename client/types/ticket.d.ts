export interface Ticket {
  id: string;
  title: string;
  price: number;
  userId: string;
  category: string;
  version: number;
}
