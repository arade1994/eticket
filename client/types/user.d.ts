import { type IncomingMessage } from "http";

export interface User {
  id: string;
  firstName: string;
  lastName: string;
  age: number;
  email: string;
  password: string;
}

export interface Rating {
  id: string;
  comment: string;
  rate: number;
  userId: string;
  ratedUserId: string;
}

export interface RequestWithUser extends IncomingMessage {
  currentUser?: { id: string; email: string };
}
