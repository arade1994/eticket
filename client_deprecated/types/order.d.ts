import { type Ticket } from "./ticket";

export interface Order {
  id: string;
  userId: string;
  status: OrderStatus;
  expiresAt: string;
  version: number;
  ticket: Ticket;
}

export declare enum OrderStatus {
  Created = "created",
  Cancelled = "cancelled",
  AwaitingPayment = "awaiting:payment",
  Complete = "complete",
}
