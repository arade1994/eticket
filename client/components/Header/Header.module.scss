.headerContainer {
  grid-column: 1/-1;
  position: sticky;
  position: -webkit-sticky;
  top: 0;

  & .navigation {
    padding: 5px 25px;
    background: linear-gradient(
      90deg,
      rgba(107, 148, 161, 1) 0%,
      rgba(7, 64, 82, 1) 100%
    );
    display: flex;
    justify-content: space-between;
    align-items: center;

    & .logo {
      font-size: 26px;
      color: white;
      text-decoration: none;
    }

    & .linksContainer {
      display: flex;
      gap: 20px;
      padding: 0 10px 0 10px;

      & a {
        color: white;
        text-decoration: none;
        border-radius: 5px;
        font-size: 20px;

        &:hover {
          color: rgb(215, 218, 217);
        }
      }

      & .activeLink {
        color: rgb(123, 206, 226);
      }
    }
  }
}
