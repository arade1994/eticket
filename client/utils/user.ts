import { Ticket } from "../types/ticket";
import { Rating, User } from "../types/user";

export const getNumOfCreatedTickets = (user: User, tickets: Ticket[]) =>
  tickets?.filter((ticket) => ticket.userId === user.id)?.length;

export const getUserRaters = (userId: string, ratings: Rating[]) => {
  return ratings.filter((rating) => rating.ratedUserId === userId);
};

export const getUserRating = (userId: string, ratings: Rating[]) => {
  const userRates = ratings
    .filter((rating) => rating.ratedUserId === userId)
    ?.map((rating) => rating.rate);
  const ratesSum = userRates.reduce((a, b) => a + b, 0);
  if (!ratesSum || isNaN(ratesSum)) return "-";
  return (ratesSum / userRates.length).toFixed(1);
};

export const getIsUserRated = (
  userId: string,
  ratings: Rating[],
  currentUserId: string
) =>
  ratings.filter(
    (rating) => rating.ratedUserId === userId && rating.userId === currentUserId
  )?.length;

export const isRatingDisabled = (
  userId: string,
  currentUserId: string,
  ratings: Rating[]
) =>
  userId === currentUserId ||
  getIsUserRated(userId, ratings, currentUserId) !== 0;

export const getUserByUserId = (userId: string, users: User[]) =>
  users.find((user) => user.id === userId);

export const filterUsers = (users: User[], searchText: string) => {
  let filteredUsers = [...users];

  if (searchText.trim() !== "")
    filteredUsers = filteredUsers.filter(
      (user) =>
        user.firstName.toLowerCase().includes(searchText) ||
        user.lastName.toLowerCase().includes(searchText)
    );

  return filteredUsers;
};
