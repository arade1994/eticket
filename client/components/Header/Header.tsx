import Link from "next/link";
import { usePathname } from "next/navigation";

import { type User } from "../../types/user";
import ProfileIcon from "../ProfileIcon/ProfileIcon";

import classes from "./Header.module.scss";

interface Props {
  currentUser?: User;
}

export default ({ currentUser }: Props) => {
  const pathname = usePathname();

  console.log(currentUser);

  return (
    <header className={classes.headerContainer}>
      <nav className={classes.navigation}>
        <Link className={classes.logo} href="/">
          eTicket
        </Link>
        {currentUser ? (
          <ProfileIcon
            firstName={currentUser.firstName}
            lastName={currentUser.lastName}
          />
        ) : (
          <div className={classes.linksContainer}>
            <Link
              className={pathname === "/auth/signin" ? classes.activeLink : ""}
              href="/auth/signin"
            >
              Sign In
            </Link>
            <Link
              className={pathname === "/auth/signup" ? classes.activeLink : ""}
              href="/auth/signup"
            >
              Sign Up
            </Link>
          </div>
        )}
      </nav>
    </header>
  );
};
