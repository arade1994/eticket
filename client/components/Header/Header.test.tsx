import { afterEach, describe, expect, test } from "vitest";
import { cleanup, render, screen } from "@testing-library/react";

import Header from "./Header";
import { User } from "../../types/user";

const mockCurrentUser = {
  id: "userId",
  firstName: "Test",
  lastName: "Test",
  age: 24,
  email: "test@test.com",
  password: "test",
} as User;

describe("<Header />", () => {
  afterEach(() => cleanup());

  test("It should match a snapshot when user is logged out", () => {
    const { baseElement } = render(<Header />);

    expect(baseElement).toMatchSnapshot("Logged out user");
  });

  test("It should display specific navigation items when user is logged out", () => {
    render(<Header />);

    const navigationLinks = screen.getAllByRole("link");
    expect(navigationLinks).toBeDefined();
    expect(navigationLinks.length).toEqual(3);
    expect(navigationLinks[0].textContent).toEqual("eTicket");
    expect(navigationLinks[1].textContent).toEqual("Sign In");
    expect(navigationLinks[2].textContent).toEqual("Sign Up");
  });

  test("It should match a snapshot when user is logged in", () => {
    const { baseElement } = render(<Header currentUser={mockCurrentUser} />);

    expect(baseElement).toMatchSnapshot("Logged in user");
  });

  test("It should display specific navigation items when user is logged in", async () => {
    render(<Header currentUser={mockCurrentUser} />);

    const navigationLinks = screen.getAllByRole("link");
    expect(navigationLinks).toBeDefined();
    expect(navigationLinks.length).toEqual(1);
    expect(navigationLinks[0].textContent).toEqual("eTicket");
  });
});
