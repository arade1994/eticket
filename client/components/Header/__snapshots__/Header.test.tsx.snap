// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Header /> > It should match a snapshot when user is logged in > Logged in user 1`] = `
<body>
  <div>
    <header
      class="_headerContainer_0a265d"
    >
      <nav
        class="_navigation_0a265d"
      >
        <a
          class="_logo_0a265d"
          href="/"
        >
          eTicket
        </a>
        <div
          class="_profileIcon_a7b94c"
        >
          <p
            style="margin: auto;"
          >
            TT
          </p>
        </div>
      </nav>
    </header>
  </div>
</body>
`;

exports[`<Header /> > It should match a snapshot when user is logged out > Logged out user 1`] = `
<body>
  <div>
    <header
      class="_headerContainer_0a265d"
    >
      <nav
        class="_navigation_0a265d"
      >
        <a
          class="_logo_0a265d"
          href="/"
        >
          eTicket
        </a>
        <div
          class="_linksContainer_0a265d"
        >
          <a
            class=""
            href="/auth/signin"
          >
            Sign In
          </a>
          <a
            class=""
            href="/auth/signup"
          >
            Sign Up
          </a>
        </div>
      </nav>
    </header>
  </div>
</body>
`;
